### PR TITLE
Add build pipeline for node images

### PR DESCRIPTION
**Pull request by NymanRobin**
The base of this comes from dev-tools repositories packer build. This has been modified to use DIB

Metal3-dev-env test cluster is tested with both Ubuntu and Centos images

### CODE REVIEW

File: jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+set -eux
+
+KUBERNETES_VERSION=${KUBERNETES_VERSION:-"v1.29.0"}
+KUBERNETES_MINOR_VERSION=${KUBERNETES_VERSION%.*}
+FILESYSTEM_TYPE="${FILESYSTEM_TYPE:-$FILESYSTEM_TYPE_DEFAULT}"
```
Comment (tuminoid) [inline jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos:7]:
```suggestion
FILESYSTEM_TYPE="${FILESYSTEM_TYPE:-${FILESYSTEM_TYPE_DEFAULT}}"
```

File: jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+OS=CentOS_9_Stream
+ROOT_PASSWORD="${ROOT_PASSWORD:-$ROOT_PASSWORD_DEFAULT}"
```
Comment (tuminoid) [inline jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos:9]:
```suggestion
ROOT_PASSWORD="${ROOT_PASSWORD:-${ROOT_PASSWORD_DEFAULT}}"
```

File: jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+VERSION=v1.29
```
Comment (tuminoid) [inline jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos:10]:
```suggestion
CRIO_VERSION=v1.29
```

File: jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+[kubernetes]
+name=Kubernetes
+baseurl=https://pkgs.k8s.io/core:/stable:/${KUBERNETES_MINOR_VERSION}/rpm/
+enabled=1
+gpgcheck=1
+gpgkey=https://pkgs.k8s.io/core:/stable:/${KUBERNETES_MINOR_VERSION}/rpm/repodata/repomd.xml.key
+exclude=kubelet kubeadm kubectl cri-tools kubernetes-cni
+EOF
+
+curl -L -o /etc/yum.repos.d/devel:kubic:libcontainers:stable.repo https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/$OS/devel:kubic:libcontainers:stable.repo
```
Comment (tuminoid) [inline jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos:25]:
```suggestion
curl -L -o /etc/yum.repos.d/devel:kubic:libcontainers:stable.repo "https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/${OS}/devel:kubic:libcontainers:stable.repo"
```

File: jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+curl -L -o /etc/yum.repos.d/devel:kubic:libcontainers:stable:cri-o:$VERSION.repo https://download.opensuse.org/repositories/devel:kubic:libcontainers:stable:cri-o:$VERSION/$OS/devel:kubic:libcontainers:stable:cri-o:$VERSION.repo
```
Comment (tuminoid) [inline jenkins/image_building/dib_elements/centos-node/pre-install.d/55-setup-repos:26]:
```suggestion
curl -L -o "/etc/yum.repos.d/devel:kubic:libcontainers:stable:cri-o:${CRIO_VERSION}.repo" "https://download.opensuse.org/repositories/devel:kubic:libcontainers:stable:cri-o:${VERSION}/${OS}/devel:kubic:libcontainers:stable:cri-o:${CRIO_VERSION}.repo"
```
What a horrible URL :)

File: jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh
Comment (tuminoid) [inline jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh:1]: Can you please properly format this file to look like other scripts in the PR.

File: jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+set -e
+url="${1}"
+dst="${2}"
+filename="$(basename "${url}")"
+tmpfile="/tmp/${filename}"
+curl -sSL -w "%{http_code}" "${url}" | sed "s:/usr/bin:/usr/local/bin:g" > /tmp/"${filename}"
```
Comment (tuminoid) [inline jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh:7]: `--fail` is your friend here, all the DIY error handling here seems unnecessary.

File: jenkins/image_building/build-node-image.sh [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+set -eux
+
+delete_old_images() {
+
+  # We keep last RETENTION_NUM of metal3ci images and delete old ones when new image has been pushed
+  # Example name for metal3ci image: metal3-ci-ubuntu-20240306T0701Z-7b988cb
+  RETENTION_NUM=5
+
+  # Delete outdated old images and keep last RETENTION_NUM
+  mapfile -t < <(openstack image list  -f json | \
```
Comment (tuminoid) [inline jenkins/image_building/build-node-image.sh:12]:
```suggestion
  mapfile -t < <(openstack image list -f json | \
```

File: jenkins/image_building/build-node-image.sh [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    echo "${MAPFILE[i]} has been deleted!"
+  done
+}
+
+export IMAGE_OS="${IMAGE_OS}"
+
+current_dir="$(dirname "$(readlink -f "${0}")")"
+
+# Disable needrestart interactive mode
+sudo sed -i "s/^#\$nrconf{restart} = 'i';/\$nrconf{restart} = 'a';/" /etc/needrestart/needrestart.conf   > /dev/null
```
Comment (tuminoid) [inline jenkins/image_building/build-node-image.sh:30]:
```suggestion
sudo sed -i "s/^#\$nrconf{restart} = 'i';/\$nrconf{restart} = 'a';/" /etc/needrestart/needrestart.conf > /dev/null
```

**Review by tuminoid — COMMENTED** [review]
Some early nitting.

**Comment by NymanRobin** [comment]
Thanks @tuminoid for the comments I will address them!

File: jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh
Comment (NymanRobin) [reply jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh:1]: Sure, but I would like first to know if this script is used and where in that case.

I see this script is also redefined in two other places in:
metal3-dev-env/tests/roles/run_tests/templates/cluster-template-workers-kubeadm-config-centos.yaml
metal3-dev-env/tests/roles/run_tests/templates/cluster-template-controlplane-kubeadm-config-centos.yaml

I wasn't able to locate any instances of its usage with a quick search. It appears that @kashifest may be the original author. Could you provide some insight into if it's required? If it's no longer necessary, we can remove it from all instances. Alternatively, if it's still in use, we can streamline the script and minimize code duplication.

Comment (kashifest) [reply jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh:1]: I dont recall it either, lets remove it and see if it breaks anything, anyhow we have it in dev-env so removing it shouldn't be an issue.

**Comment by mboukhalfa** [comment]
@NymanRobin check the workflow here branches are created on the nordix clone

**Comment by NymanRobin** [comment]
> @NymanRobin check the workflow here https://github.com/Nordix/metal3-dev-tools/blob/main/wow/github-workflow.md branches are created on the nordix fork

@mboukhalfa Thanks for letting me know I can open the PR there instead, however in the link you sent me it clearly says the step is optional :)

**Comment by NymanRobin** [comment]
I now got the images tested and cleaned this up a bit, however, the build should still be tested from Jenkins before proceeding and for that I would require CI access. Further the CI files for the ci and node image building should be combined to reduce the code a bit, but will do this after I can test it out

File: jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh
Comment (NymanRobin) [reply jenkins/image_building/node-image-cloud-init/retrieve.configuration.files.sh:1]: Removed it!

**Review by tuminoid — COMMENTED** [review]
/approve

/cc @Sunnatillo @kashifest
PTAL

**Review by kashifest — COMMENTED** [review]
/approve

**Comment by tuminoid** [comment]
/hold
until test passes, tide is going nuts in the background.

**Pull request merged**